### PR TITLE
Add changes to identify external traffic for infosec data publishing

### PR DESCRIPTION
**Pull request by imesh94**
## Add changes to identify external traffic for infosec data publishing

> Due to the multiple redirects happening in the /authorize call and internal traffic to endpoints like /authorize and /token,  metrics data becomes inaccurate. This PR contains changes to identify external calls before publishing infosec endpoint data using a header that can be added to the external requests by the load balancer. Repetitive data publishing due to /authorize redirects is also stopped by checking for sessionDataKey query parameter.

> Removing APIM side proxies defined for infosec endpoints will be sent in a separate PR.

**Issue link:** *https://github.com/wso2/financial-services-accelerator/issues/157*

**Doc Issue:** *Optional, link issue from [documentation repository](https://github.com/wso2/docs-ob/issues)*

**Applicable Labels:** *Spec, product, version, type (specify requested labels)*

------

### Development Checklist

1. [ ] Built complete solution with pull request in place.
2. [ ] Ran checkstyle plugin with pull request in place.
3. [ ] Ran Findbugs plugin with pull request in place.
4. [ ] Ran FindSecurityBugs plugin and verified report.
5. [ ] Formatted code according to WSO2 code style.
6. [ ] Have you verify the PR does't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [ ] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)?

### Testing Checklist

1. [ ] Written unit tests.
2. [ ] Documented test scenarios(link available in guides).
3. [ ] Written automation tests (link available in guides).
4. [ ] Verified tests in multiple database environments (if applicable).
5. [ ] Verified tests in multiple deployed specifications (if applicable).
6. [ ] Tested with OBBI enabled  (if applicable).
7. [ ] Tested with specification regulatory conformance suites  (if applicable).

**Automation Test Details**

| Test Suite        | Test Script IDs   |
| ----------------- | ----------------- |
| Integration Suite | *TCXXXXX, TCXXXX* |

**Conformance Tests Details**

| Test Suite Name  | Test Suite Version | Scenarios  | Result   |
| ---------------- | ------------------ | ---------- | -------- |
| *Security Suite* | *VX.X*             | *Foo, Bar* | *Passed* |

## Resources

**Knowledge Base:** https://sites.google.com/wso2.com/open-banking/

**Guides:** https://sites.google.com/wso2.com/open-banking/developer-guides

### CODE REVIEW

File: components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/InfoSecDataPublishingFilter.java [hunk 1]
```diff
@@ -245,4 +255,15 @@ private String extractClientId(HttpServletRequest request) {
...
 
+    /**
+     * Check whether data should be published for the current request.
+     *
+     * @return boolean
+     */
+    public boolean shouldPublishCurrentRequestData(ServletRequest request) {
+
+        // Check if current request is external traffic
+        return expectedExternalTrafficHeaderValue.equals(
```
Comment (nirmal070125) [inline components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/InfoSecDataPublishingFilter.java:266]: should we do `equalsIgnoreCase`?

File: components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/AuthorizeDataPublishingFilter.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * Authorize Data Publishing Filter.
+ * Implements custom logic related to publishing /authorize request data.
+ */
+public class AuthorizeDataPublishingFilter extends InfoSecDataPublishingFilter {
+
+    @Override
+    public boolean shouldPublishCurrentRequestData(ServletRequest request) {
+
+        // Check if the sessionDataKey is available in the request to prevent publishing the internal redirects.
+        return request.getAttribute("sessionDataKey") != null && super.shouldPublishCurrentRequestData(request);
```
Comment (nirmal070125) [inline components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/AuthorizeDataPublishingFilter.java:32]: There's no constant to refer to this attribute? `sessionDataKey`

**Review by nirmal070125 — COMMENTED** [review]
Is it possible to write a test case to handle the end-to-end scenario by mocking out the components?

**Comment by imesh94** [comment]
> Is it possible to write a test case to handle the end-to-end scenario by mocking out the components?

Added in https://github.com/wso2/reference-implementation-consumerdatastandards-au/pull/24/commits/81aef69b832cc2c50d286c4317b9449d10ddbe18

File: components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/InfoSecDataPublishingFilter.java (hunk 1 above)
Comment (imesh94) [reply components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/InfoSecDataPublishingFilter.java]: Fixed in https://github.com/wso2/reference-implementation-consumerdatastandards-au/pull/24/commits/5bd8bf2502b8c24f989883b4562306f313181d37

File: components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/AuthorizeDataPublishingFilter.java (hunk 2 above)
Comment (imesh94) [reply components/org.wso2.openbanking.cds.identity/src/main/java/org/wso2/openbanking/cds/identity/filter/AuthorizeDataPublishingFilter.java:32]: This is the only usage of "sessionDataKey" in java code.

**Review by aka4rKO — APPROVED** [review]

**Review by anjuchamantha — APPROVED** [review]

**Pull request merged**